### DISCUSSION TITLE
Accessors on translatable attributes no longer returning json values

### DISCUSSION THREAD

**Discussion by blackjak231** [spatie/laravel-translatable#445] [Q&A]
Hey ! 

I've been using a custom Trait mentioned [here](https://github.com/spatie/laravel-translatable/issues/320#issuecomment-1017410024) for 2 years now without any problems. To summarize, I have the following Trait with mutators : 

```
trait TranslateMethods
{
    public function getTranslatedAttribute($value)
    {
        if ($json = json_decode($value)) {
            return $json->{app()->getLocale()} ?? "";
        }
        return $value;
    }
}
``` 

```
public $translatable = [
    'title'
];

// ...

protected function title(): Attribute
{
    return Attribute::make(
        get: fn ($value) => $this->getTranslatedAttribute($value),
    );
}
```

So far, `$value` has always been equal to the json/array value of the attribute. **However**, since upgrading to Laravel 11, it seems the Attribute or event the Casting logic has changed and $value goes through each translation string in the array individually.

Here is an example of what I mean : 
Let's imaging a title attribute with the following value (in DB) : 
```
{
    'fr': 'Mon titre',
    'en': 'My title'
}
```
and the following mutator 
```
protected function title(): Attribute
{
    return Attribute::make(
        get: fn ($value) => Log::info($value),
    );
}
```
And a simple controller method : 
```
public function index(Request $request)
{
    $pages = Page::ordered()->get();
    return  response()->json(compact('pages'));
}
```

**On Laravel 10**, `Log::info` would only log `{'fr': 'Mon titre','en': 'My title'}`. If i don't log the value and us the `getTranslatedAttribute` method it returns only the translated title of the current Locale.

**On Laravel 11**, it logs `Mon titre` and then another log `My title`. If i don't log the value and us the `getTranslatedAttribute` method it returns the full json. 

I've been trying to go deeper into this library and into the way Laravel casting/mutaitng works but I've reached my max understanding after a few hours of digging...

Is anyone able to reproduce this and able to help me wrap my head around this issue ? What changed and could anything be fixed in the package ?

Thanks ! 

**FYI: I'm aware I could always do the following :**
```
protected function title(): Attribute
{
    return Attribute::make(
        get: fn ($value,$attributes) => $this->getTranslatedAttribute($attributes['title']),
    );
}
```
However I want to limit the use of static array keys.

**Comment by Tarpsvo** [comment]
https://github.com/spatie/laravel-translatable/pull/437/files

This PR and release also broke a lot of things on my side.

For example, this code:

```php
public function getImportantNoticeAttribute()
{
    if ($this->important_notice_expires?->isPast()) {
        return null;
     }

    return $this->getTranslations('important_notice');
}
```
Worked very well before release 6.6.2. Now it returns a double-nested value like this:

![image](https://github.com/spatie/laravel-translatable/assets/2018660/ecb8ba28-6180-4f02-921a-3a484c2fc0f2)

I'm not sure of what issue the PR was supposed to fix, I only feel the things that it broke.


The "fix" would be to apply it like so:
```php
public function getImportantNoticeAttribute($value)
{
    if ($this->important_notice_expires?->isPast()) {
        return null;
    }

    return $value;
}
```

But the fact that it's looped through by some magical array in the background makes no sense to me and feels counter intuitive. Not to mention that it's a breaking change. It should definitely not be a minor release.

What do you think @freekmurze?

**Reply by goaround** [reply]
I have the same problem? I use the get...Attribute function to return just the translation I need but its now always an array due to the mutateAttributeForArray function from this PR https://github.com/spatie/laravel-translatable/pull/437/files

**Reply by goaround** [reply]
At least I discovered that https://spatie.be/docs/laravel-translatable/v6/advanced-usage/customize-the-toarray-method it the better method for me.

**Comment by vencelkatai** [comment]
Hi!

I'm the author of the mentioned PR.
It supposed to fix an inconsistency when using accessors between simple property access and array casts (`Model::toArray()`).

I think the package works as expected by checking your examples. If you'd like to keep only the current locale translations in the array representation, the preferred way should be [https://spatie.be/docs/laravel-translatable/v6/advanced-usage/customize-the-toarray-method](https://spatie.be/docs/laravel-translatable/v6/advanced-usage/customize-the-toarray-method) like @goaround mentioned.

@blackjak231 You could just delete the accessor altogether and customize the toArray method instead.

@Tarpsvo In your case I think the implementation of the accessor is incorrect. It shouldn't return all the translations, using `$model->important_notice`  you should get it translated to the current locale, like translated attributes without accessors behave.
That "fix" should be the correct implementation. in my opinion it only broke because of this. It was broken before, you can check the [PR](https://github.com/spatie/laravel-translatable/pull/437) for some details if you're interested.

Let me know if I can help with anything else!
